openedx/tests/xblock_integration/xblock_testcase.py

```diff
@@ -0,0 +1,202 @@
+        )
+
+        cls.chapter = ItemFactory.create(
+            parent=cls.course,
+            display_name='Overview',
+            category='chapter'
+        )
+
+        cls.urls = {}
+
+        with cls.store.bulk_operations(cls.course.id, emit_signals=False):
+            for section in cls.test_configuration:
+                section_item = ItemFactory.create(
+                    parent=cls.chapter,
+                    display_name=section['urlname'],
+                    category='sequential'
+                )
+
+                cls.urls[section['urlname']] = reverse(
+                    'courseware_section',
+                    kwargs={
+                        'course_id': unicode(cls.course.id),
+                        'chapter': 'Overview',
+                        'section': section['urlname'],
+                    }
+                )
+                print section['urlname']
+                print cls.urls[section['urlname']]
+
+                blocks = section["xblocks"]
+
+                unit_item = ItemFactory.create(
+                    parent=section_item,
+                    display_name="unit_"+section['urlname'],
+                    category='vertical'
+                )
+                for block in blocks:
+                    block_item = ItemFactory.create(
+                        parent=unit_item,
+                        category=block['blocktype'],
+                        display_name=block['urlname']
+                    )
+            cls.course_url = reverse(
+                'courseware_section',
+                kwargs={
+                    'course_id': unicode(cls.course.id),
+                    'chapter': 'Overview',
+                    'section': section['urlname'],
+                }
+            )
+            print cls.course_url
+
+            cls.course_url = reverse(
+                'courseware_section',
+                kwargs={
+                    'course_id': unicode(cls.course.id),
+                    'chapter': u'Overview',
+                    'section': u'two_done_block_test',
+                }
+            )
+            print cls.course_url
+
+    def _containing_section(self, block_urlname):
+        '''
+        For a given block, return the parent section
+        '''
+        for section in self.test_configuration:
+            blocks = section["xblocks"]
+            for block in blocks:
+                if block['urlname'] == block_urlname:
+                    return section['urlname']
+        raise Exception("Block not found "+block_urlname)
+
+    def assertXBlockScreenshot(self, block_urlname, rendering=None):
+        '''
+        As in Bok Choi, but instead of a CSS selector, we pass a
+        block_id. We may want to be able to pass an optional selector
+        for picking a subelement of the block.
+
+        This confirms status code, and that the screenshot is
+        identical.
+
+        TODO: IMPLEMENT
+        '''
+        pass
```

Comment:
This absolutely is fragile. It is also testing for *the* major failure case with XBlocks. 

bok_choy tests are not that helpful for most XBlocks. The JavaScript is usually "Hello World"-level simple, and the only failure mode is if e.g. JQuery is missing or similar. I have never seen a reversion or failure there. 

What is fragile is the CSS/styling. Virtually all XBlocks I wrote more than a year ago are unusable in edx-platform. 

This is what the ProfileXBlock looks like today: 

![selection_013](https://cloud.githubusercontent.com/assets/1427775/11658338/123756d2-9d90-11e5-9f54-fdfe350e6267.png)

Unless you were to test every piece of css on every div, you could not catch things like this. It breaks in a different way with each XBlock. DoneXBlock, the vertical spacing became wonky. Etc.

It seems reasonable that when you install a new XBlock and run tests, it confirms it renders in a sensible way in your runtime. When you change styling, the test suite shows you all of the blocks which now look different, and you confirm that nothing insane happened. 

When we were discussing the issue @benpatterson pointed me to: http://bok-choy.readthedocs.org/en/latest/visual_diff.html and I thought something like this would be very sensible to have.